app/Http/Controllers/Api/BookController.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace App\Http\Controllers\Api;
+
+use Illuminate\Http\Request;
+use App\Http\Controllers\Controller;
+use DB;
+use App\Model\Book;
+use App\Model\Post;
+use App\Model\Borrowing;
+use App\Model\QrCode;
+use App\Model\Rating;
+use App\Model\Comment;
+use App\Model\Favorite;
+
+class BookController extends Controller
+{
+    /**
+     * Soft delete "book" and its relationship ("borrowing", "post", "qrcode", "comment"),
+     * Hard delete "rating" with id of book.
+     *
+     * @param Request $request request
+     * @param int     $id      id of book
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy(Request $request, $id)
+    {
+        DB::beginTransaction();
+        try {
+            $book = Book::find($id)->delete();
+            DB::commit();
+        } catch (\PDOException $e) {
+            DB::rollBack();
+        }
+        if ($request->ajax()) {
+            return response()->json(['book'=> $book], 200);
+        }
+    }
+
+    /**
+     * Restore book and its relationship with id of book.
+     *
+     * @param Request $request request
+     * @param int     $id      id of book
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function restore(Request $request, $id)
+    {
+        $timeDelete = Book::withTrashed()->where('id', $id)->pluck('deleted_at')->first();
+        DB::beginTransaction();
+        try {
+            //Restore book and favorites of its.
+            $book = Book::withTrashed()->find($id)->restore();
+            Favorite::withTrashed()->where('deleted_at', $timeDelete)->where('favoritable_type', Favorite::TYPE_BOOK)->where('favoritable_id', $id)->restore();
+
+            //Restore rating, qrcode, borrowing.
+            Rating::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            QrCode::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            Borrowing::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+
+            //Restore post. Get list post was restored. Restored all comment and favorites for each post.
+            $posts = Post::withTrashed()->select('id')->where('book_id', $id)->where('deleted_at', $timeDelete)->get();
+            Post::withTrashed()->where('book_id', $id)->where('deleted_at', $timeDelete)->restore();
+            foreach ($posts as $post) {
```

Comment:
- you should get array posts id and use whereIn to restore. In this case, you should don't use foreach statement.